django/src/games/pong.py

```diff
@@ -0,0 +1,272 @@
+                await self.send_message("group", args=data)
+        except ValueError as e:
+            print(f"Value error: {e}")
+            await self.send_error(str(e))
+        except Exception as e:
+            print(f"Unexpected error in receive: {e}")
+            await self.send_error("An unexpected error occurred")
+
+    async def game_join(self, event):
+        if self.host:
+            self.info.players.append(event["content"]["user"])
+            if len(self.info.players) == self.info.player_needed:
+                await self.redis.set(self.room_id, 1)
+            await self.send_message("group", "game_info", self.info.__dict__)
+
+    async def game_info(self, event):
+        await self.send_message("client", args=event)
+
+    async def game_ready(self, event):
+        if self.host:
+            self.info.player_ready += 1
+            if self.info.player_ready == self.info.player_needed:
+                await self.send_message("group", "game_start")
+
+    async def game_start(self, event):
+        await self.send_message("client", args=event)
+        if self.host:
+            self.game_task = asyncio.create_task(self.loop())
+
+    async def game_state(self, event):
+        await self.send_message("client", args=event)
+
+    async def game_move(self, event):
+        if self.host:
+            await self.move_pad(event["content"]["pad_n"], event["content"]["direction"])
+
+    async def game_stop(self, event):
+        if hasattr(self, "offline"):
+            return
+        if self.host:
+            self.winner = event["content"].get("winner")
+            if not self.winner:
+                self.punish_coward(event["content"].get("user"))
+        await self.send_message("client", args=event)
+        await self.close()
+
+    def punish_coward(self, coward):
+        if not coward:
+            return
+        if coward is self.info.players[0]:
+            self.info.score[1] = self.info.score[0] + 1
+        else:
+            self.info.score[0] = self.info.score[1] + 1
+        self.winner = self.get_winner()
+
+    async def loop(self):
+        win = 5
+        fps = 0.033
+        while True:
+            try:
+                self.ball.move()
+                self.handle_collisions()
+                if self.info.score[0] == win or self.info.score[1] == win:
+                    await self.send_message("group", "game_stop", {"winner": self.get_winner()})
+                    break
+                await self.send_game_state()
+                await asyncio.sleep(fps)
+            except Exception as e:
+                print(f"Unexpected error in the loop from game {self.room_id}: {e}")
+                break
+
+    def get_winner(self):
+        if self.mode == "online":
+            return self.info.players[0] if self.info.score[0] > self.info.score[1] else self.info.players[1]
+        else:
+            return "Pad 1" if self.info.score[0] > self.info.score[1] else "Pad 2"
```

Comment:
subject require all players to have usernames

Reply:
It's only for local games